File: server/src/controllers/dashboardController.ts
import { Request, Response } from 'express';
import { query } from '../config/db.js';
import moment from 'moment';

export const getDashboardStats = async (req: Request, res: Response) => {
  try {
    // Get all data from database
    const [projects, personnel, materials] = await Promise.all([
      query<any[]>('SELECT * FROM projects'),
      query<any[]>('SELECT * FROM personnel'),
      query<any[]>('SELECT * FROM materials'),
    ]);

    // Calculate stats
    const activeProjects = projects.filter((p) => p.status === 'in_progress');
    const totalRevenue = projects.reduce((sum: number, p: any) => sum + parseFloat(p.actual_cost || 0), 0);
    const totalExpenses = totalRevenue * 0.85; // Can be calculated from actual expenses
    const totalProfit = totalRevenue - totalExpenses;

    const activePersonnel = personnel.filter((p) => p.status === 'active');

    const lowStockMaterials = materials.filter(
      (m) => parseFloat(m.current_stock || 0) <= parseFloat(m.min_stock || 0)
    );

    // Recent projects (last 5)
    const recentProjects = projects
      .sort((a, b) => moment(b.created_at).valueOf() - moment(a.created_at).valueOf())
      .slice(0, 5)
      .map((p) => ({
        id: p.id,
        name: p.name,
        status: p.status,
        progress: p.progress || 0,
        createdAt: p.created_at,
      }));

    res.json({
      totalProjects: projects.length,
      activeProjects: activeProjects.length,
      totalRevenue,
      totalExpenses,
      totalProfit,
      totalPersonnel: personnel.length,
      activePersonnel: activePersonnel.length,
      lowStockMaterials: lowStockMaterials.length,
      recentProjects,
    });
  } catch (error: any) {
    console.error('Error fetching dashboard stats:', error);
    res.status(500).json({ error: 'Không thể lấy thống kê dashboard' });
  }
};

File: server.net/Controllers/DashboardController.cs
using ConstructOs.Server.Config;
using Microsoft.AspNetCore.Mvc;

namespace ConstructOs.Server.Controllers;

[ApiController]
[Route("api/dashboard")]
public class DashboardController : ControllerBase
{
    private readonly IDatabaseService _db;

    public DashboardController(IDatabaseService db)
    {
        _db = db;
    }

    [HttpGet("stats")]
    public async Task<IActionResult> GetDashboardStats()
    {
        try
        {
            var projects = await _db.QueryAsync<dynamic>("SELECT * FROM projects");
            var personnel = await _db.QueryAsync<dynamic>("SELECT * FROM users");
            var materials = await _db.QueryAsync<dynamic>("SELECT * FROM materials");

            var activeProjects = projects.Where(p => p.status?.ToString() == "in_progress").ToList();
            var totalRevenue = projects.Sum(p => Convert.ToDouble(p.actual_cost ?? 0));
            var totalExpenses = totalRevenue * 0.85;
            var totalProfit = totalRevenue - totalExpenses;

            var activePersonnel = personnel.Where(p => p.status?.ToString() == "active").ToList();

            var lowStockMaterials = materials.Where(m =>
            {
                var currentStock = Convert.ToDouble(m.current_stock ?? 0);
                var minStock = Convert.ToDouble(m.min_stock ?? 0);
                return currentStock <= minStock;
            }).ToList();

            var recentProjects = projects
                .OrderByDescending(p => DateTime.Parse(p.created_at?.ToString() ?? DateTime.MinValue.ToString()))
                .Take(5)
                .Select(p => new
                {
                    id = p.id,
                    name = p.name,
                    status = p.status,
                    progress = p.progress ?? 0,
                    createdAt = p.created_at
                })
                .ToList();

            return Ok(new
            {
                totalProjects = projects.Count(),
                activeProjects = activeProjects.Count,
                totalRevenue,
                totalExpenses,
                totalProfit,
                totalPersonnel = personnel.Count(),
                activePersonnel = activePersonnel.Count,
                lowStockMaterials = lowStockMaterials.Count,
                recentProjects
            });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error fetching dashboard stats: {ex}");
            return StatusCode(500, new { error = "Không thể lấy thống kê dashboard" });
        }
    }
}

